pkg/controller/builder.go

```diff
@@ -247,6 +252,62 @@ func UpdateDeploymentProxyHook(pc clients.ProxyClientInterface) deploymentcontro
 		if len(errs) > 0 {
 			return errors.Join(errs...)
 		}
+
+		return nil
+	}
+}
+
+func UpdateDeploymentFeatureGatesHook(
+	featuresAccessor featuregates.FeatureGateAccess,
+	featuresMapper internalfeatures.MapperInterface,
+) deploymentcontroller.DeploymentHookFunc {
+	return func(_ *operatorv1.OperatorSpec, deployment *appsv1.Deployment) error {
+		klog.FromContext(context.Background()).WithName("builder").V(0).Info("FeatureGatesHook updating environment", "deployment", deployment.Name)
+
+		featuregates, err := featuresAccessor.CurrentFeatureGates()
+		if err != nil {
+			return fmt.Errorf("error getting featuregates.config.openshift.io/cluster: %w", err)
+		}
+
+		var argVal string
+		if featuregates.Enabled(features.FeatureGateNewOLM) {
+			upstreamGates, err := featuresMapper.UpstreamForDownstream(features.FeatureGateNewOLM)
+			if err != nil {
+				return fmt.Errorf("error getting upstream feature gates mapping: %w", err)
+			}
+			argVal = internalfeatures.FormatAsEnabledArgs(upstreamGates)
+		}
+
+		var errs []error
+		for i := range deployment.Spec.Template.Spec.Containers {
+			klog.FromContext(context.Background()).WithName("builder").V(4).Info("FeatureGatesHook iterating containers", "container", deployment.Spec.Template.Spec.Containers[i].Name)
+			if !strings.EqualFold(deployment.Spec.Template.Spec.Containers[i].Name, "manager") {
+				continue
+			}
```

Comment:
This is saying "only add the feature gates flag if the container name is manager"? If so, +1. I think that is a reasonable convention. And it's one of the answers I had in mind for my earlier question about dealing with the question of which containers to inject this flag into.

Reply:
yes, that was the idea here, but with the both components having separate sets of feature flags, I had to include deployment name in the mix, see https://github.com/openshift/cluster-olm-operator/pull/102#discussion_r1930344434